### PR TITLE
Release script [WIP]

### PR DESCRIPTION
**Pull request by tlrobinson**
This doesn't need to be run within a Metabase repo since it pulls it from git. It probably won't hurt though, aside from polluting your repo with `metabase` and `metabase-deploy` directories. Alternatively, in another directory you could do `../metabase/bin/release` etc.
1. [x] Update `bin/version`, commits, tags and pushes.
2. [x] Build the jar
3. [x] Sign the jar
4. [x] Push jar to S3
5. [x] Build and push Docker image
6. [x] Build and push EB config to S3
7. [x] Push the Heroku app (`metabase-deploy`) 
8. [x] Tag Docker image as `latest` (`docker pull metabase/metabase:vX.Y.Z && docker tag metabase/metabase:vX.Y.Z metabase/metabase:latest && docker push metabase/metabase:latest`)
9. [ ] Build/push Mac app
10. [ ] Update homebrew recipe
11. [ ] Build/deploy docs/website
12. [ ] Publish Github release
13. [ ] Publish blog post
14. [ ] Update `version-info.json`
15. [ ] Close GitHub milestone
16. [ ] Merge `release-X.Y.Z` back into `master`

I'd also like to split this up into smaller scripts so we can run a subset of the tasks.

### CODE REVIEW

**Comment by salsakran** [comment]
hmm. @tlrobinson + @camsaul, I'm thinking of adding to this but think I want to add to in with python. Specifically, I'm looking at creating something that auto-generates release notes from tags.

thoughts?

**Comment by tlrobinson** [comment]
I don't think we should add another language to the mix. Node.js can almost certainly do what you want. If you give me details I can do it. Or write it in Python and I'll port it (assuming it's not large)

**Comment by camsaul** [comment]
I think honestly all of our build scripts should be in python. It's probably the best tool for the job here

**Comment by salsakran** [comment]
@tlrobinson + @camsaul added some changes here.

File: bin/release [hunk 1]
```diff
@@ -109,16 +114,16 @@ bin/aws-eb-docker/release-eb-version.sh "v$VERSION"
...
+echo "pulling"
 cd metabase-deploy
-git fetch
-git co "origin/master"
+git pull
 
+echo "release heroku artifacts"
 bin/release-heroku "v$VERSION"
-
-# technologic
```
Comment (salsakran) [inline bin/release]: cause I'm sick of side-chained kick drums ?

Comment (camsaul) [inline bin/release]: side-chaining low-pass filters to the kicks is still the business. how would you do trance music with no side-chaining ⛓

File: bin/release [hunk 2]
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env bash
+
+set -eu
+
+DEFAULT_KEYSTORE_PATH="$PWD/certs/metabase_keystore.jks"
+DEFAULT_KEY_PEM_PATH="$PWD/certs/key.pem"
+DEFAULT_CONFIG_JSON_PATH="$PWD/config.json"
```
Comment (camsaul) [inline bin/release]: so these things are all `DEFAULT`, but is there any way to configure them?

File: bin/release [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+# ensure we have access to the keystore
+if [ -f "$DEFAULT_KEYSTORE_PATH" ]; then
+    KEYSTORE_PATH="$DEFAULT_KEYSTORE_PATH"
+else
+    echo "Enter the path to JAR signing keystore ($DEFAULT_KEYSTORE_PATH by default)"
+    read KEYSTORE_PATH
+    [ -f "$KEYSTORE_PATH" ] || exit 1
+fi
+
+# # ensure we've configured the Mac signing key
```
Comment (camsaul) [inline bin/release]: If you're going to check in a bunch of commented-out code I would say at least add a comment explaining why it's commented out

**Comment by camsaul** [comment]
I would say the second half looks good, it explains what's going on. Not sure I'm following the tasks in the first half.

**Comment by tlrobinson** [comment]
Here's a script to validate the Docker tags.

```bash
#!/usr/bin/env bash

set -eu

validate-docker-tag() {
  tag="$1"
  expected_version="$2"
  docker pull "metabase/metabase:$tag"
  actual_version="$(docker run --rm -it --entrypoint java metabase/metabase:$tag -jar /app/metabase.jar version | grep -oE 'v\d+\.\d+\.\d+[^, ]*')"
  if [ "$expected_version" != "$actual_version" ]; then
    echo "Docker tag '$tag' has incorrect version '$actual_version', expected '$expected_version'";
    return 1
  else
    echo "Docker tag '$tag' has correct version '$actual_version'";
    return 0
  fi
}

validate-docker-tag "v0.22.0" "v0.22.0"
validate-docker-tag "latest" "v0.22.0"
```

**Review by camsaul — APPROVED** [review]

**Comment by kdoh** [comment]
@metabase/core-developers What do we need to do to close this out?

**Comment by kdoh** [comment]
Going to go ahead and close this to reflect the fact that this is not being actively developed. I'd love for us to finish this, or do something like it, but since it's failing tests and hasn't been touched in around 5 months, I don't want to mislead folks.

**Pull request closed without merging**